code2flow/php.py

```diff
@@ -0,0 +1,399 @@
+
+        :param tree ast:
+        :returns: tuple of group, node, and body trees. These are processed
+                  downstream into real Groups and Nodes.
+        :rtype: (list[ast], list[ast], list[ast])
+        """
+        tree = tree or []  # if its abstract, it comes in with no body
+
+        groups = []
+        nodes = []
+        body = []
+        for el in tree:
+            if el['nodeType'] in ('Stmt_Function', 'Stmt_ClassMethod', 'Expr_Closure'):
+                nodes.append(el)
+            elif el['nodeType'] in ('Stmt_Class', 'Stmt_Namespace', 'Stmt_Trait'):
+                groups.append(el)
+            else:
+                tup = PHP.separate_namespaces(children(el))
+                if tup[0] or tup[1]:
+                    groups += tup[0]
+                    nodes += tup[1]
+                    body += tup[2]
+                else:
+                    body.append(el)
+        return groups, nodes, body
+
+    @staticmethod
+    def make_nodes(tree, parent):
+        """
+        Given a tree element of all the lines in a function, create the node along
+        with the calls and variables internal to it.
+        Also make the nested subnodes
+
+        :param tree ast:
+        :param parent Group:
+        :rtype: list[Node]
+        """
+        assert tree['nodeType'] in ('Stmt_Function', 'Stmt_ClassMethod', 'Expr_Closure')
+
+        if tree['nodeType'] == 'Expr_Closure':
+            token = '(Closure)'
+        else:
+            token = tree['name']['name']
+        is_constructor = token == '__construct' and parent.group_type == GROUP_TYPE.CLASS
+
+        tree_body = tree['stmts']
+        subgroup_trees, subnode_trees, this_scope_body = PHP.separate_namespaces(tree_body)
+        assert not subgroup_trees
+        calls = make_calls(this_scope_body)
+        variables = make_local_variables(this_scope_body, parent)
+
+        if parent.group_type == GROUP_TYPE.CLASS and parent.parent.group_type == GROUP_TYPE.NAMESPACE:
+            import_tokens = [djoin(parent.parent.token, parent.token, token)]
+        if parent.group_type in (GROUP_TYPE.NAMESPACE, GROUP_TYPE.CLASS):
+            import_tokens = [djoin(parent.token, token)]
+        else:
+            import_tokens = [token]
+
+        node = Node(token, calls, variables, parent, import_tokens=import_tokens,
+                    is_constructor=is_constructor, line_number=lineno(tree))
+
+        subnodes = flatten([PHP.make_nodes(t, parent) for t in subnode_trees])
+        return [node] + subnodes
+
+    @staticmethod
+    def make_root_node(lines, parent):
+        """
+        The "root_node" is an implict node of lines which are executed in the global
+        scope on the file itself and not otherwise part of any function.
+
+        :param lines list[ast]:
+        :param parent Group:
+        :rtype: Node
+        """
+        token = "(global)"
+        line_number = lines and lineno(lines[0]) or 0
```

Comment:
```suggestion
        line_number = lineno(lines[0]) if lines else 0
```

This seems to be a bit tricky. Can you use a ternary operator instead?

Reply:
Yup. Ternary is a better choice.